processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -267,55 +291,160 @@ static double estimatePredicateSelectivity(
   }
 
   static double estimatePredicateSelectivity(
+      BitmapIndex bitmapIndex,
       ColumnSelector columnSelector,
       String dimension,
-      Iterable<ImmutableBitmap> bitmaps,
+      Iterable<Integer> bitmapIndexes,
       long totalNumRows
   )
   {
     final ColumnCapabilities columnCapabilities = columnSelector.getColumn(dimension).getCapabilities();
-    return estimatePredicateSelectivity(
-        bitmaps,
+    return estimateSelectivityOfBitmapList(
+        bitmapIndex,
+        bitmapIndexes,
         totalNumRows,
+        // assume multi-value column if columnCapabilities is null
         columnCapabilities == null || columnCapabilities.hasMultipleValues()
     );
   }
 
-  static double estimatePredicateSelectivity(
+  private static double estimateSelectivityOfBitmapList(
+      BitmapIndex bitmapIndex,
+      Iterable<Integer> bitmapIndexeIterable,
+      long totalNumRows,
+      boolean isMultiValueDimension
+  )
+  {
+    final List<Integer> bitmapIndexes = ImmutableList.copyOf(bitmapIndexeIterable);
+    long numMatchedRows = 0;
+    for (Integer index : bitmapIndexes) {
+      final ImmutableBitmap bitmap = bitmapIndex.getBitmap(index);
+      numMatchedRows += bitmap.size();
+    }
+
+    if (isMultiValueDimension) {
+      final double estimated = numMatchedRows * computeNonOverlapRatioFromRandomBitmapSamples(
+          bitmapIndex,
+          bitmapIndexes
+      ) / totalNumRows;
+      return Math.min(1., estimated);
+    } else {
+      return (double) numMatchedRows / totalNumRows;
+    }
+  }
+
+  static double estimateSelectivityOfBitmapTree(
       Iterable<ImmutableBitmap> bitmaps,
       long totalNumRows,
       boolean isMultiValueDimension
   )
   {
     long numMatchedRows = 0;
-    final List<ImmutableBitmap> bitmapList = Lists.newArrayList(bitmaps);
-    for (ImmutableBitmap bitmap : bitmapList) {
+    for (ImmutableBitmap bitmap : bitmaps) {
       numMatchedRows += bitmap.size();
     }
 
-    // assume multi-value column if columnCapabilities is null
     if (isMultiValueDimension) {
-      final double estimated = numMatchedRows * Filters.computeNonOverlapRatioFromBitmapSamples(bitmapList)
-                               / totalNumRows;
+      final double estimated = numMatchedRows * computeNonOverlapRatioFromFirstNBitmapSamples(
```

Comment:
Oops. I'm such an idiot. It's my bad. I'll change this method to calculate the average non-overlap ratio of the right hand side bitmap of the union operation like below.
(The below is just an idea, and not optimized)

```
while (...) {
  overlapSize = size(unionedBitmap) + size(rightBitmap) - size(union(unionedBitmap, rightBitmap))
  unionedBitmap = union(unionedBitmap, rightBitmap)
  nonOverlapRatioOfRightBitmap = (size(rightBitmap) - overlapSize) / rightBitmap.size()
  nonOverlapRatioSum += nonOverlapRatioOfRightBitmap
}
return nonOverlapRatioSum / sampleNum;
```

And, ```estimateSelectivityOfBitmapList()``` calling this method will be changed like below.

```
double numMatchedRows = bitmapIndex.getBitmap(bitmapIndexes.get(0)).size();
for (int i = 1; i < bitmapIndexes.size(); i++) {
  final ImmutableBitmap bitmap = bitmapIndex.getBitmap(bitmapIndexes.get(i));
  numMatchedRows += bitmap.size() * nonOverlapRatio;
}
return numMatchedRows / totalNumRows
```

I'll do add some tests.